packages/grid/x-data-grid/src/hooks/features/columnHeaders/useGridColumnHeaders.tsx

```diff
@@ -268,20 +300,215 @@ export const useGridColumnHeaders = (props: UseGridColumnHeadersProps) => {
       );
     }
 
+    return (
+      <GridColumnHeaderRow role="row" aria-rowindex={headerGroupingMaxDepth + 1}>
+        {columns}
+      </GridColumnHeaderRow>
+    );
+  };
+
+  const getColumnsHeaderGroups = (params?: {
+    renderContext: GridRenderContext | null;
+    minFirstColumn?: number;
+    maxLastColumn?: number;
+  }) => {
+    if (headerGroupingMaxDepth === 0) {
+      return null;
+    }
+
+    const {
+      renderContext: nextRenderContext = renderContext,
+      minFirstColumn = minColumnIndex,
+      maxLastColumn = visibleColumns.length,
+    } = params || {};
+
+    if (!nextRenderContext) {
+      return null;
+    }
+
+    const columns: JSX.Element[] = [];
+
+    const [firstRowToRender, lastRowToRender] = getRenderableIndexes({
+      firstIndex: nextRenderContext.firstRowIndex,
+      lastIndex: nextRenderContext.lastRowIndex,
+      minFirstIndex: 0,
+      maxLastIndex: currentPage.rows.length,
+      buffer: rootProps.rowBuffer,
+    });
+
+    const firstColumnToRender = getFirstColumnIndexToRenderRef.current({
+      firstColumnIndex: nextRenderContext!.firstColumnIndex,
+      minColumnIndex: minFirstColumn,
+      columnBuffer: rootProps.columnBuffer,
+      apiRef,
+      firstRowToRender,
+      lastRowToRender,
+      visibleRows: currentPage.rows,
+    });
+
+    const lastColumnToRender = Math.min(
+      nextRenderContext.lastColumnIndex! + rootProps.columnBuffer,
+      maxLastColumn,
+    );
+
+    const renderedColumns = visibleColumns.slice(firstColumnToRender, lastColumnToRender);
```

Comment:
This block is the same as in `getColumnsHeader`. We could extract it to a function is reuse in both.